datafusion/datasource-parquet/src/file_format.rs

```diff
@@ -557,38 +559,175 @@ pub fn coerce_int96_to_resolution(
     file_schema: &Schema,
     time_unit: &TimeUnit,
 ) -> Option<Schema> {
-    let mut transform = false;
-    let parquet_fields: HashMap<_, _> = parquet_schema
+    // Traverse the parquet_schema columns looking for int96 physical types. If encountered, insert
+    // the field's full path into a set.
+    let int96_fields: HashSet<_> = parquet_schema
         .columns()
         .iter()
-        .map(|f| {
-            let dt = f.physical_type();
-            if dt.eq(&Type::INT96) {
-                transform = true;
-            }
-            (f.name(), dt)
-        })
+        .filter(|f| f.physical_type().eq(&Type::INT96))
+        .map(|f| f.path().string())
         .collect();
 
-    if !transform {
+    if int96_fields.is_empty() {
+        // The schema doesn't contain any int96 fields, so skip the remaining logic.
         return None;
     }
 
-    let transformed_fields: Vec<Arc<Field>> = file_schema
-        .fields
-        .iter()
-        .map(|field| match parquet_fields.get(field.name().as_str()) {
-            Some(Type::INT96) => {
-                field_with_new_type(field, DataType::Timestamp(*time_unit, None))
+    // Do a DFS into the schema using a stack, looking for timestamp(nanos) fields that originated
+    // as int96 to coerce to the provided time_unit.
+
+    type NestedFields = Rc<RefCell<Vec<FieldRef>>>;
+    type StackContext<'a> = (
+        Vec<&'a str>, // The full parquet path to the field currently being processed.
```

Comment:
Sorry @mbutrovich please help me to understand that, why the parquet paths are needed? 
My feeling the method works on the schema level rather than file path level

Reply:
By parquet path I mean the parquet column path like "c0.list.element.c1".

Reply:
Ic, make sense, perhaps we can rename it to `fully qualified column path` or `dot notation path` something like that?